lib/select.c

```diff
@@ -151,8 +163,20 @@ int Curl_select(curl_socket_t maxfd,   /* highest socket number */
     ptimeout = NULL;
   }
   else if(timeout_ms > 0) {
-    pending_tv.tv_sec = timeout_ms / 1000;
-    pending_tv.tv_usec = (timeout_ms % 1000) * 1000;
+    pending_tv.tv_sec = (long)(timeout_ms / 1000);
+    pending_tv.tv_usec = (long)((timeout_ms % 1000) * 1000);
+#ifdef WIN32
+    /* tv_sec overflow check on Windows there it is still long */
+    if(pending_tv.tv_sec < 0)
+      pending_tv.tv_sec = LONG_MAX;
+#elif TIME_T_MIN < 0
+    /* tv_sec overflow check in case time_t is signed */
+    if(pending_tv.tv_sec < 0)
+      pending_tv.tv_sec = TIME_T_MAX;
+#endif
+    /* tv_usec overflow check since it must support -1...1000000 */
+    if(pending_tv.tv_usec < 0)
+      pending_tv.tv_usec = 999999;
```

Comment:
Does usec *need* to support negative values? For POSIX that seems to be true: *"The type suseconds_t shall be a signed integer type capable of storing values at least in the range [-1, 1000000]."* ([ref](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/types.h.html#tag_13_67)). Not sure we should count on all systems we run on to be that strictly POSIX complient.

Perhaps more importantly, it is wrong to assume that we will get a negative value when we typecast it to a smaller variable type, as signed integer overflow in C is not defined behavior. The overflow must be detected *before* it happens.

Reply:
Thanks. Checks are still done after the calculation, but the calculation is still done with `timediff_t` as suggested by @jay and then converted afterwards.